src/main/java/org/opentripplanner/transit/model/network/TripPattern.java

```diff
@@ -42,6 +42,11 @@
  * stop). Trips are assumed to be non-overtaking, so that an earlier trip never arrives after a
  * later trip.
  * <p>
+ * The Route and StopPattern together form the natural key of the TripPattern. They are the shared
+ * set of characteristics that group many trips together into one TripPattern. This grouping saves
+ * memory by not replicating any details shared across all trips in the TripPattern, but it is
+ * also essential to some optimizations in routing algorithms like Raptor.
+ * <p>
```

Comment:
To the extent that TripPattern originally existed only to optimize routing, maybe it has no natural key because it's not part of a "business domain" outside of routing algorithms. The business practice may be to group trips by modes, and it might be convenient to keep these mode-groups separate e.g. to ensure variety in results. There may be a problem here where a domain model entity is being conflated with an internal routing optimization grouping. It seems to me that the key is subjective design choice and not really natural.

Whatever the key is, in my opinion there does need to be a very clear statement in this class's documentation about exactly which characteristics associate a trip with a specific TripPattern. When reading this class, it was not clear to me which characteristics were guaranteed to be identical across all trips. I will remove the word "natural" and just describe all the fields that make up the key, including the ones you've mentioned here (mode, submode).

Side note: I've come to the conclusion several times over the years that it's not necessarily a good thing to support overtaking. Because even if the data are accurate and one bus is going to overtake the other on the same pattern, would we ever give someone the advice to not board the first arriving bus, wait and take the second one? Would any rider ever follow that advice even if we gave it to them, for buses with all the exact same characteristics visiting exactly the same stops? Basically no one believes that vehicles on the same TripPattern will overtake in a reliable way. It might be a good choice to state that trips on the same pattern are allowed to overtake, but the algorithm may choose one as if they don't overtake.

Reply:
@abyrd I agree to the above comment, all three paragraphs.